### PR TITLE
Another batch of type annotations

### PR DESCRIPTION
**Pull request by DMRobertson**
Like #12716. Recommend commit-by-commit review.

`mypy -p synapse --disallow-untyped-errors` now reports 99 errors on my machine. Before the previous PR, it was around 150.

### CODE REVIEW

File: synapse/types.py [hunk 1]
```diff
@@ -852,25 +859,28 @@ def __bool__(self) -> bool:
...
-    if len(keys) != 1:
-        raise ValueError("Invalid key")
-    # and return that one key
-    for key_id, key_data in keys.items():
+    # and that it contains exactly one key
+    if len(keys) == 1:
+        key_id, key_data = keys.popitem()
         return key_id, decode_verify_key_bytes(key_id, decode_base64(key_data))
+    else:
+        raise ValueError("Invalid key")
```
Comment (DMRobertson) [inline synapse/types.py:883]: This is meant to be equivalent. Mypy doesn't like the previous version: it can't see that the loop on -872 will always run exactly once.

File: synapse/http/proxyagent.py [hunk 2]
```diff
@@ -245,7 +245,7 @@ def http_proxy_endpoint(
     proxy: Optional[bytes],
     reactor: IReactorCore,
     tls_options_factory: Optional[IPolicyForHTTPS],
-    **kwargs,
+    **kwargs: object,
```
Comment (DMRobertson) [inline synapse/http/proxyagent.py:248]: bit of a cop-out.

File: synapse/logging/scopecontextmanager.py [hunk 3]
```diff
@@ -107,36 +109,48 @@ class _LogContextScope(Scope):
...
         """
         Args:
-            manager (LogContextScopeManager):
+            manager:
                 the manager that is responsible for this scope.
             span (Span):
                 the opentracing span which this scope represents the local
                 lifetime for.
             logcontext (LogContext):
                 the logcontext to which this scope is attached.
```
Comment (DMRobertson) [inline synapse/logging/scopecontextmanager.py:128]: wasn't sure where these types came from, so this is just a drive-by.

File: synapse/logging/_remote.py [hunk 4]
```diff
@@ -118,7 +122,7 @@ def __init__(
         if _reactor is None:
             from twisted.internet import reactor
 
-            _reactor = reactor
+            _reactor = reactor  # type: ignore[assignment]
```
Comment (DMRobertson) [inline synapse/logging/_remote.py:125]: `twisted.internet.reactor` is a module that, upon import, [removes itself from `sys.modules`](
https://github.com/twisted/twisted/blob/trunk/src/twisted/internet/reactor.py#L37-L40) and calls `twisted.internet.default.install()`. This eventually [reinserts `twisted.internet.reactor` back into `sys.modules`](https://github.com/twisted/twisted/blob/d0a8ca7f28f296feefd0695acadf8dd729acfd65/src/twisted/internet/main.py#L20-L34), but now it points to a reactor instance.

Comment (DMRobertson) [reply synapse/logging/_remote.py:125]: No wonder mypy is confused.

File: synapse/logging/formatter.py [hunk 5]
```diff
@@ -28,10 +30,14 @@ class LogFormatter(logging.Formatter):
     where it was caught are logged).
     """
 
-    def __init__(self, *args, **kwargs):
-        super().__init__(*args, **kwargs)
```
Comment (DMRobertson) [inline synapse/logging/formatter.py:32]: Removed this since it did literally nothing other than create a new stack frame!

File: synapse/logging/_remote.py (hunk 4 above)
Comment (clokep) [reply synapse/logging/_remote.py:125]: In other spots we do:

```python
from twisted.internet import reactor as _reactor

reactor = cast(ISynapseReactor, _reactor)
```

🤷  (Could probably just put a `cast` here instead of an ignore to be consistent.)

File: synapse/types.py (hunk 1 above)
Comment (DMRobertson) [reply synapse/types.py:883]: This appears to have broken a test!

Comment (DMRobertson) [reply synapse/types.py:883]: Ahh. If `keys` is empty, then before this change we would return `None`. Now we explicitly error.

Either the comments were incorrect, or else the implementation was buggy.

Comment (DMRobertson) [reply synapse/types.py:883]: This hasn't changed since its initial implementation! https://github.com/matrix-org/synapse/commit/c659b9f94fff29adfb2abe4f6b345710b65e8741#diff-d5643d4a94ee928bdd677a4ffbfbf3c571bc0a9f84d00d92dde31d91a6d53533R483-R501

Comment (DMRobertson) [reply synapse/types.py:883]: Every call site that I can see expects this function to return a 2-tuple.

**Comment by DMRobertson** [comment]
Not sure if that is a sytest flake or a genuine regression. I'll re-run.

**Review by erikjohnston — APPROVED** [review]
